File: ruby/test/supermarket_test.rb
require_relative './test_helper'

class SupermarketTest < Minitest::Test
  include Approvals

  cover "Kata*"

  def setup
    @catalog = FakeCatalog.new
    @teller = Kata::Teller.new(@catalog)
    @the_cart = Kata::ShoppingCart.new

    @toothbrush = Kata::Product.new("toothbrush", Kata::ProductUnit::EACH)
    @catalog.add_product(@toothbrush, 0.99)
    @rice = Kata::Product.new("rice", Kata::ProductUnit::EACH)
    @catalog.add_product(@rice, 2.99)
    @apples = Kata::Product.new("apples", Kata::ProductUnit::KILO)
    @catalog.add_product(@apples, 1.99)
    @cherry_tomatoes = Kata::Product.new("cherry tomato box", Kata::ProductUnit::EACH)
    @catalog.add_product(@cherry_tomatoes, 0.69)
  end

  def test_an_empty_shopping_cart_should_cost_nothing
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_one_normal_item
    @the_cart.add_item(@toothbrush)
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_two_normal_items
    @the_cart.add_item(@toothbrush)
    @the_cart.add_item(@rice)
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_buy_two_get_one_free
    @the_cart.add_item(@toothbrush)
    @the_cart.add_item(@toothbrush)
    @the_cart.add_item(@toothbrush)
    @teller.add_special_offer(Kata::SpecialOfferType::THREE_FOR_TWO, @toothbrush, @catalog.unit_price(@toothbrush))
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_buy_five_get_one_free
    @the_cart.add_item(@toothbrush)
    @the_cart.add_item(@toothbrush)
    @the_cart.add_item(@toothbrush)
    @the_cart.add_item(@toothbrush)
    @the_cart.add_item(@toothbrush)
    @teller.add_special_offer(Kata::SpecialOfferType::THREE_FOR_TWO, @toothbrush, @catalog.unit_price(@toothbrush))
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_loose_weight_product
    @the_cart.add_item_quantity(@apples, 0.5)
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_percent_discount
    @the_cart.add_item(@rice)
    @teller.add_special_offer(Kata::SpecialOfferType::TEN_PERCENT_DISCOUNT, @rice, 10.0)
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_x_for_y_discount
    @the_cart.add_item(@cherry_tomatoes)
    @the_cart.add_item(@cherry_tomatoes)
    @teller.add_special_offer(Kata::SpecialOfferType::TWO_FOR_AMOUNT, @cherry_tomatoes, 0.99)
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_five_for_y_discount
    @the_cart.add_item_quantity(@apples, 5)
    @teller.add_special_offer(Kata::SpecialOfferType::FIVE_FOR_AMOUNT, @apples, 6.99)
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_five_for_y_discount_with_six
    @the_cart.add_item_quantity(@apples, 6)
    @teller.add_special_offer(Kata::SpecialOfferType::FIVE_FOR_AMOUNT, @apples, 6.99)
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_five_for_y_discount_with_sixteen
    @the_cart.add_item_quantity(@apples, 16)
    @teller.add_special_offer(Kata::SpecialOfferType::FIVE_FOR_AMOUNT, @apples, 6.99)
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

  def test_five_for_y_discount_with_four
    @the_cart.add_item_quantity(@apples, 4)
    @teller.add_special_offer(Kata::SpecialOfferType::FIVE_FOR_AMOUNT, @apples, 6.99)
    receipt = @teller.checks_out_articles_from(@the_cart)
    verify Kata::ReceiptPrinter.new(40).print_receipt(receipt)
  end

end

File: cpp/test/SuperMarketTest.cpp


#include "ApprovalTests.hpp"
#include "Catch.hpp"

#include "../model/ReceiptPrinter.h"
#include "../model/SupermarketCatalog.h"
#include "FakeCatalog.h"
#include "../model/ShoppingCart.h"
#include "../model/Teller.h"

TEST_CASE("Discounts", "[Supermarket]")
{
    SupermarketCatalog *catalog = new FakeCatalog();
    Product toothbrush("toothbrush", ProductUnit::Each);
    catalog->addProduct(toothbrush, 0.99);
    Product rice("rice", ProductUnit::Each);
    catalog->addProduct(rice, 2.99);
    Product apples("apples", ProductUnit::Kilo);
    catalog->addProduct(apples, 1.99);
    Product cherryTomatoes("cherry tomato box", ProductUnit::Each);
    catalog->addProduct(cherryTomatoes, 0.69);

    ShoppingCart cart;
    Teller teller(catalog);
    ReceiptPrinter printer = ReceiptPrinter();

    SECTION("Empty cart costs nothing")
    {
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("One normal item")
    {
        cart.addItem(toothbrush);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("Two normal items")
    {
        cart.addItem(toothbrush);
        cart.addItem(rice);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("buy two get one free")
    {
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        teller.addSpecialOffer(SpecialOfferType::ThreeForTwo, toothbrush, catalog->getUnitPrice(toothbrush));
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("buy two get one free but insufficient in basket")
    {
        cart.addItem(toothbrush);
        teller.addSpecialOffer(SpecialOfferType::ThreeForTwo, toothbrush, catalog->getUnitPrice(toothbrush));
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }
    SECTION("buy five get one free")
    {
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        cart.addItem(toothbrush);
        teller.addSpecialOffer(SpecialOfferType::ThreeForTwo, toothbrush, catalog->getUnitPrice(toothbrush));
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }


    SECTION("Loose weight product")
    {
        apples.setQuantityType("KG");
        cart.addItemQuantity(apples, 0.5);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));

    }
    SECTION("percent discount")
    {
        cart.addItem(rice);
        teller.addSpecialOffer(SpecialOfferType::TenPercentDiscount, rice, 10.0);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("x for y discount")
    {
        cart.addItem(cherryTomatoes);
        cart.addItem(cherryTomatoes);
        teller.addSpecialOffer(SpecialOfferType::TwoForAmount, cherryTomatoes, 0.99);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("x for y discount but insufficient in basket")
    {
        cart.addItem(cherryTomatoes);
        teller.addSpecialOffer(SpecialOfferType::TwoForAmount, cherryTomatoes, 0.99);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("5 for y discount")
    {
        cart.addItemQuantity(apples, 5);
        teller.addSpecialOffer(SpecialOfferType::FiveForAmount, apples, 6.99);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("5 for y discount with 6")
    {
        cart.addItemQuantity(apples, 6);
        teller.addSpecialOffer(SpecialOfferType::FiveForAmount, apples, 5.99);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("5 for y discount with 16")
    {
        cart.addItemQuantity(apples, 16);
        teller.addSpecialOffer(SpecialOfferType::FiveForAmount, apples, 7.99);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }

    SECTION("5 for y discount with 4")
    {
        cart.addItemQuantity(apples, 4);
        teller.addSpecialOffer(SpecialOfferType::FiveForAmount, apples, 8.99);
        Receipt receipt = teller.checksOutArticlesFrom(cart);
        ApprovalTests::Approvals::verify(printer.printReceipt(receipt));
    }
}
